cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/dataHandler.ts

```diff
@@ -0,0 +1,390 @@
+// which is meant to add children or assign a value depending on its type
+function jsonRec(d, data, json) {
+  data.map((k) => {
+    if (k.parentId === d.id) {
+      json[k.name] = {};
+      if (k.type === 'record' && !k.path) {
+        jsonRec(k, data, json[k.name]);
+        return;
+      }
+      json[k.name] = k.path;
+    }
+  });
+  return json;
+}
+
+// The function receives json format and is meant to convert it into the
+// correct flat schema format that the frontend will use to render the records
+export function jsonToFlat(json) {
+  let result = [];
+  const keys = Object.keys(json);
+  keys.map((key) => {
+    const uniqueId = uuidV4();
+
+    if (typeof json[key] === 'object') {
+      result.push({ id: uniqueId, name: key, parentId: null, type: 'record' });
+      result = iterateChildren(uniqueId, json[key], result);
+    }
+  });
+  return result;
+}
+
+// This function receives a parentId, and object and an array,
+// it iterates through the keys of the objects and returns a flat row
+function iterateChildren(parentId, tree, result) {
+  const keys = Object.keys(tree);
+
+  keys.map((key: string) => {
+    const uniqueId = uuidV4();
+    if (!Array.isArray(tree[key])) {
+      result.push({ id: uniqueId, name: key, parentId, type: 'record' });
+      iterateChildren(uniqueId, tree[key], result);
+      return;
+    }
+    result.push({ id: uniqueId, name: key, parentId, type: 'string', path: tree[key] });
+  });
+  return result;
+}
+
+// This function receives two arrays, one to push and one to loop, a unique id and a margin number,
+// it returns an array with flat rows
+export const handleChildren = (childrenOfSelected, item, uniqueId, m) => {
+  let marginLeft = m;
+  marginLeft += 6;
+  item.forEach((child) => {
+    const childId = uuidV4();
+    let typeOfChild: any = {};
+    typeOfChild = typeHandler(child);
+    childrenOfSelected.push({
+      path: child.path,
+      id: childId,
+      parentId: uniqueId,
+      name: child.name ? child.name : child,
+      type: typeOfChild.type,
+      children: [],
+      margin: marginLeft,
+    });
+
+    if (typeof child.type === 'object') {
+      handleChildren(childrenOfSelected, typeOfChild.children, childId, marginLeft);
+    }
+  });
+  return childrenOfSelected;
+};
+
+// This function receives the input schema fields and converts it into an array of flat rows which the UI expects
+export const inputSchemaToFlat = (fieldValues) => {
+  let dropdownData = [];
+  fieldValues.forEach((item) => {
+    let childrenOfSelected = [];
+    const uniqueId = uuidV4();
+
+    if (item.children.length !== 0) {
+      const marginLeft = 0;
+      childrenOfSelected = handleChildren(childrenOfSelected, item.children, uniqueId, marginLeft);
+    }
+    const childToBeAdded = {
+      id: uniqueId,
+      parentId: null,
+      name: item.name,
+      children: [],
+      type: item.type,
+      margin: 0,
```

Comment:
Again, while flattening input schema to flat fields we shouldn't need to store the margin. This can be computed based on the ancestors of any specific field.